File: advent_of_code/2024/day7/zig/src/part1.zig
const std = @import("std");

fn is_solvable(target: i64, numbers: []const i64) !bool {
    // Initialize a stack as an ArrayList
    const allocator = std.heap.page_allocator;
    var stack = std.ArrayList(i64).init(allocator);
    defer stack.deinit(); // Free memory at the end

    // Push all numbers to the stack
    for (numbers) |number| {
        try stack.append(number);
    }

    // Simplify the stack by adding the top two numbers repeatedly
    while (stack.items.len > 1) {
        const a = stack.pop();
        const b = stack.pop();
        const result = a + b;
        try stack.append(result);
    }

    // Check if the final result matches the target
    return stack.items[0] == target;
}

pub fn solve(input: []const u8) !i64 {
    const allocator = std.heap.page_allocator;
    var sum: i64 = 0;

    // Split input string by comma
    var it = std.mem.splitScalar(u8, input, ',');
    const target_str = it.next() orelse return error.InvalidInput;
    const numbers_str = it.next() orelse return error.InvalidInput;

    // Parse the target number
    const target = try std.fmt.parseInt(i64, target_str, 10);

    // Split numbers_str by spaces and parse them into an ArrayList
    var numbers = std.ArrayList(i64).init(allocator);
    defer numbers.deinit(); // Free memory at the end

    var numbers_it = std.mem.splitScalar(u8, numbers_str, ' ');
    while (numbers_it.next()) |number_str| {
        if (number_str.len > 0) { // Ignore empty parts
            const number = try std.fmt.parseInt(i64, number_str, 10);
            try numbers.append(number);
        }
    }

    // Check if the target can be achieved
    if (try is_solvable(target, numbers.items)) {
        sum += target;
    }

    return sum;
}

File: advent_of_code/2024/day7/rust/src/part1.rs
fn is_solvable(
    target: i64,
    numbers: &Vec<i64>,
    curr_idx: usize,
    curr_num: i64,
    sequence: &mut Vec<char>,
) -> (bool, Vec<char>) {
    if curr_idx == numbers.len() {
        if curr_num == target {
            return (true, sequence.clone());
        } else {
            return (false, vec![]);
        }
    }

    let next_num = numbers[curr_idx];
    sequence.push('+');
    let is_plus_solv = is_solvable(target, numbers, curr_idx + 1, curr_num + next_num, sequence);
    sequence.pop();
    sequence.push('*');
    let is_mult_solv = is_solvable(target, numbers, curr_idx + 1, curr_num * next_num, sequence);
    sequence.pop();

    if is_plus_solv.0 {
        is_plus_solv
    } else if is_mult_solv.0 {
        is_mult_solv
    } else {
        (false, vec![])
    }
}

pub fn solve(input: &str) -> i64 {
    let mut sum = 0i64;
    for line in input.lines() {
        let (left, right) = line.split_once(": ").unwrap();
        let target = left.trim().parse::<i64>().unwrap();
        let operands = right
            .split(" ")
            .map(|x| x.trim().parse::<i64>().unwrap())
            .collect::<Vec<i64>>();

        let is_solv = is_solvable(target, &operands, 1, operands[0], &mut vec![]);
        if is_solv.0 {
            sum += target;
            println!("Target: {}", target);
            println!("Operands: {:?}", operands);
            println!("Sequence: {:?}", is_solv.1);
        }
    }

    sum
}
